### PR TITLE
[BugFix][Attention] Fix stale KV lengths in hamming sparse attention under ACL graph replay

### PR DESCRIPTION
**Pull request by grandolantow**
# [BugFix][Attention] Fix stale KV lengths in hamming sparse attention under ACL graph replay

- **Commit**: `4edfcc103d2b01b7cffded2971fa792620ab9f0f` (vllm-ascend)
- **Modules**: `vllm_ascend/attention/attention_v1.py`, `vllm_ascend/attention/kvcomp_attn/attention_utils.py`
- **Type**: BugFix (accuracy)
- **Impact scope**: Only the combination of "hamming sparse enabled + ACL full-graph replay"; all other configurations are unaffected

---

## Summary

On the ACL full-graph replay path, the KV sequence length passed to the FIA operator for hamming sparse attention was not refreshed to the hamming-reduced value. As a result, the attention computed in graph mode diverged from the eager-mode result, causing an accuracy regression. This PR adds a per-layer override that rebinds the KV length correctly.

---

## Background

During the decode phase, hamming sparse attention selects the top-k most relevant KV blocks per request via hamming distance and only attends over those blocks. The corresponding KV length is `seq_lens_from_hamming` (shorter than the true full length). It is computed in `attention_utils.py:48-52`:

```python
new_seq_lens = torch.where(
    remainder == 0,
    chunk_size * top_k_cpu,                    # evenly divisible: exactly top-k blocks
    chunk_size * (top_k_cpu - 1) + remainder,  # with remainder: append the tail
)
```

The eager path (`forward_fused_infer_attention` -> `get_kvcomp_decode_params`) has always used this reduced length and produces correct results.

---

## The Problem

Under ACL full-graph replay, when a captured FIA subgraph is replayed, tensors such as `query` / `key_cache` / `value` / `block_table` are rebound to the latest metadata, but `seq_lens` (i.e. the FIA operator's `actual_seq_lengths_kv`) is kept at the full length `seq_lens_list` captured at capture time and is **not** replaced with `seq_lens_from_hamming`.

Consequence: the replayed graph actually attends over the **full KV**, inconsistent with the eager path (which only attends over the selected top-k blocks) -> accuracy regression.

### Trigger condition

The graph-replay path is only taken when the master switch at `model_runner_v1.py:2495-2499` is true:

```python
if (cudagraph_runtime_mode == CUDAGraphMode.FULL   # using ACL full-graph mode
    and not forward_context.capturing               # currently "replaying", not "capturing"
    and not self.use_sparse and not self.use_compress):  # not the DSA compression path
```

---

### Behavior comparison

| Aspect | Before | After |
| --- | --- | --- |
| KV length passed to FIA on replay | full `seq_lens_list` | hamming-reduced `seq_lens_from_hamming` |
| Hamming-active layer | attends full KV -> diverges from eager | attends selected top-k blocks -> aligned with eager |
| skip/rollback layer | full length | full length (unchanged, skipped via `hashk_caches[i] is None`) |
| Non-hamming model | original behavior | original behavior (`hasattr` short-circuits, zero impact) |

### Worked example

Consider a decode step with 1 request that has generated `seq_len = 512`, `block_size = 128` (4 blocks in total), and hamming selects top-k = 2 most relevant blocks.

- `remainder = 512 % 128 = 0` -> `seq_lens_from_hamming = 128 x 2 = 256`

The `actual_seq_lengths_kv` received by the FIA operator is then:

```
eager path (correct baseline): 256   -> attend over 2 selected blocks
graph replay, BEFORE:          512   -> attend over all 4 blocks   <- bug
graph replay, AFTER:           256   -> attend over 2 selected blocks   <- fixed
```

> Analogy: graph replay is like "record a recipe, then swap ingredients". At capture time the "how much to use" (seq_lens) was recorded; at replay, the query/key/block_table ingredients are swapped for fresh ones, but the "how much" bowl was left unchanged -> it still uses the full amount recorded earlier. This PR swaps that bowl too, at replay time, with the hamming-reduced amount.

---

## The Fix

In the FIA replay branch (non-draft path) of `update_graph_params` (`attention_v1.py:402`), `seq_lens` is overridden per layer:

```python
obj = attn_metadata[key]
if (hasattr(obj, 'kvcomp_metadata')
        and hasattr(obj.kvcomp_metadata, 'seq_lens_from_hamming')
        and obj.kvcomp_metadata.hashk_caches[layer_count] is not None):
    seq_lens = obj.kvcomp_metadata.seq_lens_from_hamming   # active layer: reduced length
elif not hasattr(vllm_config.model_config.hf_text_config, "sliding_window"):
    block_tables = attn_metadata[key].block_tables          # non-SWA: original behavior
layer_count += 1
```

Per-layer semantics:

- **Hamming-active layer** (`hashk_caches[i]` not None) -> overridden with `seq_lens_from_hamming` ✓
- **skip/rollback layer** (`hashk_caches[i]` is None) -> condition not met, keeps full length ✓
- **Non-hamming model** -> `obj` has no `kvcomp_metadata` attribute -> first `hasattr` is False, the whole `and` short-circuits -> falls through to the original `elif` branch, behavior unchanged ✓

### Incidental change: config key alignment

`is_enable_hamming_sparse()` (`attention_utils.py:145`) migrates from the legacy flat key `enable_hamming_sparse` (bool) to the new nested key `hamming_sparse` (dict containing `enabled` and `sparse_json_location`), aligning with the format already used by `AscendConfig` (`ascend_config.py:277-280`).

```python
# before
enable_hamming_sparse = additional_config.get("enable_hamming_sparse", False)
# after
hamming_sparse = additional_config.get("hamming_sparse", {"enabled": False, "sparse_json_location": ""})
enable_hamming_sparse = hamming_sparse["enabled"]
```

---

## Call chain

```
model_runner_v1.execute_model (1663)
  -> _model_forward (2514)
    -> _update_full_graph_params_if_needed (2489)        [trigger condition above]
      -> acl_graph.update_full_graph_params (226)
        -> impl_cls.update_graph_params (attention_v1.py:402)   <- fix location
          -> for key in zip(attn_keys, attn_params, handles, events):
               non-draft branch: override seq_lens = seq_lens_from_hamming (638-639)
               -> npu_fused_infer_attention_score.out(..., actual_seq_lengths_kv=seq_lens) (658)
```

Compared with the eager path: `get_kvcomp_decode_params` (`attention_utils.py:116-142`) returns **both** the new `block_table` (`hamming_output`) and `seq_lens_from_hamming`; the graph-replay path only overrides `seq_lens`.

---

- vLLM version: v0.23.0
- vLLM main: https://github.com/vllm-project/vllm/commit/e5588e49bc2642670116664a7fc4096e27adb179

### CODE REVIEW

File: vllm_ascend/attention/attention_v1.py [hunk 1]
```diff
@@ -634,8 +635,16 @@ def extract_layer_index(key: str) -> int:
...
-                        if not hasattr(vllm_config.model_config.hf_text_config, "sliding_window"):
+                        obj = attn_metadata[key]
+                        if (
+                            hasattr(obj, "kvcomp_metadata")
+                            and hasattr(obj.kvcomp_metadata, "seq_lens_from_hamming")
+                            and obj.kvcomp_metadata.hashk_caches[layer_count] is not None
+                        ):
+                            seq_lens = obj.kvcomp_metadata.seq_lens_from_hamming
+                        elif not hasattr(vllm_config.model_config.hf_text_config, "sliding_window"):
                             block_tables = attn_metadata[key].block_tables
```
Comment (weijinqian0) [inline vllm_ascend/attention/attention_v1.py]: Plz make sure that if c8 has same bug.

File: vllm_ascend/attention/kvcomp_attn/attention_utils.py [hunk 2]
```diff
@@ -147,6 +147,7 @@ def is_enable_hamming_sparse():
     if vllm_config is None:
         return False
     additional_config = vllm_config.additional_config if vllm_config.additional_config is not None else {}
-    enable_hamming_sparse = additional_config.get("enable_hamming_sparse", False)
+    hamming_sparse = additional_config.get("hamming_sparse", {"enabled": False, "sparse_json_location": ""})
```
Comment (kunpengW-code) [inline vllm_ascend/attention/kvcomp_attn/attention_utils.py]: This modification appears unrelated to bug fixes and can be removed.

**Comment by grandolantow** [comment]
/rerun
[Bot]: rerun completed.

**Rerun:**
- E2E

**Review by weijinqian0 — APPROVED** [review]

File: vllm_ascend/attention/kvcomp_attn/attention_utils.py [hunk 3]
```diff
@@ -149,4 +149,13 @@ def is_enable_hamming_sparse():
...
     enable_hamming_sparse = enable_hamming_sparse and not vllm_config.speculative_config
+    # Defense-in-depth: AscendConfig already guards this combo at config-parse
+    # time. Mirror the guard here so any code path that reads the hamming flag
+    # directly (e.g. attention __init__) also fails fast instead of silently
+    # running C8 without the hamming hooks.
+    if enable_hamming_sparse:
+        quant_config = getattr(vllm_config, "quant_config", None)
+        assert not getattr(quant_config, "enable_c8_quant", False), (
+            "Hamming sparse attention does not support C8 KV cache quantization."
+        )
```
Comment (kunpengW-code) [inline vllm_ascend/attention/kvcomp_attn/attention_utils.py]: C8 check: retain only here, delete elsewhere

**Comment by grandolantow** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: still in progress, retry /rerun after completion

**Comment by grandolantow** [comment]
/rerun
[Bot]: rerun completed.

**Rerun:**
- E2E

**Review by weijinqian0 — APPROVED** [review]
